File: src/main/java/com/fengsheng/phase/CheckWin.java
package com.fengsheng.phase;

import com.fengsheng.Fsm;
import com.fengsheng.Game;
import com.fengsheng.Player;
import com.fengsheng.ResolveResult;
import com.fengsheng.card.Card;
import com.fengsheng.protos.Common;
import com.fengsheng.skill.SkillId;
import org.apache.log4j.Logger;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * 判断是否有人胜利
 * <li>只有接收阶段正常接收情报才会进入 {@link ReceivePhaseSenderSkill} </li>
 * <li>其它情况均为置入情报区，一律进入这里。</li>
 */
public class CheckWin implements Fsm {
    private static final Logger log = Logger.getLogger(CheckWin.class);

    /**
     * 谁的回合
     */
    public Player whoseTurn;
    /**
     * 接收第三张黑色情报的顺序
     */
    public final ReceiveOrder receiveOrder;
    /**
     * 濒死结算后的下一个动作
     */
    public Fsm afterDieResolve;

    public CheckWin(Player whoseTurn, Fsm afterDieResolve) {
        this(whoseTurn, new ReceiveOrder(), afterDieResolve);
    }

    public CheckWin(Player whoseTurn, ReceiveOrder receiveOrder, Fsm afterDieResolve) {
        this.whoseTurn = whoseTurn;
        this.receiveOrder = receiveOrder;
        this.afterDieResolve = afterDieResolve;
    }

    @Override
    public ResolveResult resolve() {
        Game game = whoseTurn.getGame();
        Player stealer = null; // 簒夺者
        List<Player> redPlayers = new ArrayList<>(), bluePlayers = new ArrayList<>();
        for (Player p : game.getPlayers()) {
            if (p.isLose()) continue;
            switch (p.getIdentity()) {
                case Black -> {
                    if (p.getSecretTask() == Common.secret_task.Stealer) stealer = p;
                }
                case Red -> redPlayers.add(p);
                case Blue -> bluePlayers.add(p);
            }
        }
        List<Player> declareWinner = new ArrayList<>(), winner = new ArrayList<>();
        boolean redWin = false, blueWin = false;
        for (Player player : game.getPlayers()) {
            if (player.isLose()) continue;
            int red = 0, blue = 0;
            for (Card card : player.getMessageCards().values()) {
                for (Common.color color : card.getColors()) {
                    switch (color) {
                        case Red -> red++;
                        case Blue -> blue++;
                    }
                }
            }
            switch (player.getIdentity()) {
                case Black:
                    if (player.getSecretTask() == Common.secret_task.Collector && (red >= 3 || blue >= 3)) {
                        declareWinner.add(player);
                        winner.add(player);
                    }
                    break;
                case Red:
                    if (red >= 3) {
                        declareWinner.add(player);
                        redWin = true;
                    }
                    break;
                case Blue:
                    if (blue >= 3) {
                        declareWinner.add(player);
                        blueWin = true;
                    }
            }
        }
        if (redWin) winner.addAll(redPlayers);
        if (blueWin) winner.addAll(bluePlayers);
        if (!declareWinner.isEmpty() && stealer != null && stealer.equals(whoseTurn)) {
            declareWinner = List.of(stealer);
            winner = new ArrayList<>(declareWinner);
        }
        if (!declareWinner.isEmpty()) {
            boolean hasGuXiaoMeng = false;
            for (Player p : winner) {
                if (p.findSkill(SkillId.WEI_SHENG) != null && p.isRoleFaceUp()) {
                    hasGuXiaoMeng = true;
                    break;
                }
            }
            if (hasGuXiaoMeng) {
                for (Player p : game.getPlayers()) {
                    if (!p.isLose() && p.getIdentity() == Common.color.Has_No_Identity) {
                        winner.add(p);
                    }
                }
            }
            var declareWinners = declareWinner.toArray(new Player[0]);
            var winners = winner.toArray(new Player[0]);
            log.info(Arrays.toString(declareWinners) + "宣告胜利，胜利者有" + Arrays.toString(winners));
            for (Player p : game.getPlayers())
                p.notifyWin(declareWinners, winners);
            whoseTurn.getGame().end();
            return new ResolveResult(null, false);
        }
        return new ResolveResult(new StartWaitForChengQing(whoseTurn, receiveOrder, afterDieResolve), true);
    }
}

File: src/main/java/phase/CheckWin.kt
package com.fengsheng.phase

import com.fengsheng.*
import com.fengsheng.phase.CheckWinimport

com.fengsheng.protos.Common.colorimport com.fengsheng.protos.Common.secret_taskimport com.fengsheng.skill.Skillimport com.fengsheng.skill.SkillIdimport java.util.List com.fengsheng.protos.Common
import java.util.concurrent.LinkedBlockingQueue
import io.netty.util.HashedWheelTimerimport

org.apache.log4j.Loggerimport java.util.*import java.util.List

/**
 * 判断是否有人胜利
 *  * 只有接收阶段正常接收情报才会进入 [ReceivePhaseSenderSkill]
 *  * 其它情况均为置入情报区，一律进入这里。
 */
class CheckWin(
    /**
     * 谁的回合
     */
    var whoseTurn: Player,
    /**
     * 接收第三张黑色情报的顺序
     */
    val receiveOrder: ReceiveOrder,
    /**
     * 濒死结算后的下一个动作
     */
    var afterDieResolve: Fsm?
) : Fsm {
    constructor(whoseTurn: Player, afterDieResolve: Fsm?) : this(whoseTurn, ReceiveOrder(), afterDieResolve)

    override fun resolve(): ResolveResult? {
        val game = whoseTurn.game
        var stealer: Player? = null // 簒夺者
        var mutator: Player? = null // 诱变者
        val redPlayers: MutableList<Player> = ArrayList()
        val bluePlayers: MutableList<Player> = ArrayList()
        for (p in game.players) {
            if (p.isLose) continue
            when (p.identity) {
                color.Black -> {
                    if (p.secretTask == secret_task.Stealer) stealer =
                        p else if (p.secretTask == secret_task.Mutator) mutator = p
                }

                color.Red -> redPlayers.add(p)
                color.Blue -> bluePlayers.add(p)
            }
        }
        var declareWinner: MutableList<Player> = ArrayList()
        var winner: MutableList<Player> = ArrayList()
        var redWin = false
        var blueWin = false
        var mutatorMayWin = false
        for (player in game.players) {
            if (player.isLose) continue
            var red = 0
            var blue = 0
            for (card in player.messageCards.values) {
                for (color in card.colors) {
                    if (color == Common.color.Red) red++ else if (color == Common.color.Blue) blue++
                }
            }
            if (red >= 3 || blue >= 3) {
                mutatorMayWin = true
            }
            when (player.identity) {
                color.Black -> {
                    if (player.secretTask == secret_task.Collector && (red >= 3 || blue >= 3)) {
                        declareWinner.add(player)
                        winner.add(player)
                    }
                }

                color.Red -> {
                    if (red >= 3) {
                        declareWinner.add(player)
                        redWin = true
                    }
                }

                color.Blue -> {
                    if (blue >= 3) {
                        declareWinner.add(player)
                        blueWin = true
                    }
                }
            }
        }
        if (redWin) {
            winner.addAll(redPlayers)
            if (game.players.size == 4) winner.addAll(bluePlayers)
        }
        if (blueWin) {
            winner.addAll(bluePlayers)
            if (game.players.size == 4) winner.addAll(redPlayers)
        }
        if (declareWinner.isEmpty() && mutator != null && mutatorMayWin) {
            declareWinner.add(mutator)
            winner.add(mutator)
        }
        if (!declareWinner.isEmpty() && stealer != null && stealer === whoseTurn) {
            declareWinner = List.of(stealer)
            winner = ArrayList(declareWinner)
        }
        if (!declareWinner.isEmpty()) {
            var hasGuXiaoMeng = false
            for (p in winner) {
                if (p.findSkill<Skill?>(SkillId.WEI_SHENG) != null && p.isRoleFaceUp) {
                    hasGuXiaoMeng = true
                    break
                }
            }
            if (hasGuXiaoMeng) {
                for (p in game.players) {
                    if (!p.isLose && p.identity == color.Has_No_Identity) {
                        winner.add(p)
                    }
                }
            }
            val declareWinners = declareWinner.toTypedArray()
            val winners = winner.toTypedArray()
            log.info(Arrays.toString(declareWinners) + "宣告胜利，胜利者有" + Arrays.toString(winners))
            game.allPlayerSetRoleFaceUp()
            for (p in game.players) p.notifyWin(declareWinners, winners)
            whoseTurn.game.end(winner)
            return ResolveResult(null, false)
        }
        return ResolveResult(StartWaitForChengQing(whoseTurn, receiveOrder, afterDieResolve), true)
    }

    companion object {
        private val log = Logger.getLogger(CheckWin::class.java)
    }
}